File: dcp-solutions/2024-11-14/solution.rb
```ruby
def spiral_matrix(matrix)
  rows, cols = matrix.size
  visited = Array::NULL * rows * cols
  directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
  row, col, dir_idx = 0, 0, 0

  while row < rows && col < cols
    (0...rows).each do |i|
      next if visited[i * cols + col]
      print matrix[i][col], " "
      visited[i * cols + col] = true
    end
    col += directions[dir_idx][1]
    next unless col < cols && !visited[row * cols + col]

    (0...cols).each do |i|
      next if visited[i * cols + row]
      print matrix[row][i], " "
      visited[i * cols + row] = true
    end
    row += directions[dir_idx][0]
    next unless row < rows && !visited[(row - 1) * cols + col]

    (cols - 1).downto(0).each do |i|
      next if visited[i * cols + row]
      print matrix[row][i], " "
      visited[i * cols + row] = true
    end
    row -= directions[dir_idx][0]
    next unless row >= 0 && !visited[(row + 1) * cols + col]

    (rows - 1).downto(0).each do |i|
      next if visited[i * cols + col]
      print matrix[i][col], " "
      visited[i * cols + col] = true
    end
    col -= directions[dir_idx][1]
    next unless col >= 0 && !visited[(row - 1) * cols + (col - 1)]

    dir_idx = (dir_idx + 1) % 4
  end

  print "\n"
end

matrix = [
  [1, 2, 3, 4, 5],
  [6, 7, 8, 9, 10],
  [11, 12, 13, 14, 15],
  [16, 17, 18, 19, 20]
]

spiral_matrix(matrix)
```

File: dcp-solutions/2024-11-14/solution.ex
```elixir
defmodule MatrixSpiral do
  def spiral(matrix) do
    {rows, cols} = Integer.sqrt(Enum.count(matrix) / 2) |> round |> <<>>,
    matrix = Enum.map(0..rows-1, fn _ -> Enum.to_list(matrix[]) end),
    directions = [[0, 1], [1, 0], [0, -1], [-1, 0]],
    directions_index = 3,
    row = 0,
    col = 0
    |>
      fn() ->
        until_row do
          print_matrix(matrix[row], row, col)
          move_direction
        end
        Enum.at(directions, directions_index) |> elem(0) |> add_to(row, col)
        |>
          fn() -> move_direction end
        end

    until_row = fn ->
      {row, _} = Enum.find_idx(matrix, &(&1 != []))
      row
    end
    move_direction =
      fn ->
        directions_index
        |> add_to(directions_index, 1)
        |> mod!(4)
      end
    print_matrix = fn matrix, row, col ->
      for i <- 0..Enum.count(matrix) - 1 do
        IO.write(inspect(Enum.at(matrix[row], i)), " ")
        Enum.at(matrix, row) = Enum.pop(Enum.at(matrix, row))
      end
      IO.write("\n")
    end
    add_to = fn a, b -> a + b end
end

defmodule MatrixSpiralTest do
  def run_test() do
    matrix =
      Enum.map(1..5, &Enum.to_list(Enum.map(1..10, fn _ -> rand() end)) )
        |> List.flatten()
        |> Enum.chunk_every(2)
    MatrixSpiral.spiral.(matrix).()
  end

end
```